src/node/eviction.cpp

```diff
@@ -238,3 +241,194 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& evicti
     // Disconnect from the network group with the most connections
     return vEvictionCandidates.front().id;
 }
+
+EvictionManagerImpl::EvictionManagerImpl() {}
+EvictionManagerImpl::~EvictionManagerImpl() = default;
+
+void EvictionManagerImpl::AddCandidate(NodeId id, std::chrono::seconds connected,
+                                       uint64_t keyed_net_group, bool prefer_evict,
+                                       bool is_local, Network network,
+                                       bool noban, ConnectionType conn_type)
+{
+    LOCK(m_candidates_mutex);
+    // The default values used to initialise the eviction candidates result in
+    // newer candidates being more likely to be evicted.
+    NodeEvictionCandidate candidate{
+        Desig(id) id,
+        Desig(m_connected) connected,
+        Desig(m_min_ping_time) std::chrono::microseconds::max(),
+        Desig(m_last_block_time) 0s,
+        Desig(m_last_tx_time) 0s,
+        Desig(fRelevantServices) false,
+        Desig(m_relay_txs) false,
+        Desig(fBloomFilter) false,
+        Desig(nKeyedNetGroup) keyed_net_group,
+        Desig(prefer_evict) prefer_evict,
+        Desig(m_is_local) is_local,
+        Desig(m_network) network,
+        Desig(m_noban) noban,
+        Desig(m_conn_type) conn_type,
+    };
+    m_candidates.emplace_hint(m_candidates.end(), id, std::move(candidate));
+}
+
+bool EvictionManagerImpl::RemoveCandidate(NodeId id)
+{
+    LOCK(m_candidates_mutex);
+    return m_candidates.erase(id) != 0;
+}
+
+std::optional<NodeId> EvictionManagerImpl::SelectNodeToEvict() const
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    {
+        LOCK(m_candidates_mutex);
+        for (const auto& [id, candidate] : m_candidates) {
+            candidates.push_back(candidate);
+        }
+    }
+    return ::SelectNodeToEvict(std::move(candidates));
+}
+
+void EvictionManagerImpl::UpdateMinPingTime(NodeId id, std::chrono::microseconds ping_time)
```

Comment:
/me idly wonders whether these setters can be templated to avoid repetitive boilerplate code

Reply:
Tried to come up with a cool way of doing this with templates but didn't end with anything useful.

Reply:
I think something like this would work:

```c++
class EvictionManagerImpl
{
    ...
    template <typename Fn>
    auto CandidateOp(NodeId id, Fn&& fn)
    {
        LOCK(m_candidates_mutex);
        auto it = m_candidates.find(offset);
        if (it != m_candidates.end()) {
            return fn(it->second);
        } else {
            // return a dummy value of the right type, including void
            return ([]() -> decltype(fn(it->second)) { return {}; })();
        }
    }
    ...
};

void EvictionManager::UpdateLastBlockTime(NodeId id, std::chrono::seconds block_time)
{
    m_impl->CandidateOp(id, [](NodeEvictionCandidate& nec) {
        nec.m_last_block_time = block_time;
    });
}

void EvictionManager::GetLastBlockTime(NodeId id) const
{
    return m_impl->CandidateOp(id, [](const NodeEvictionCandidate& nec) {
        return nec.m_last_block_time;
    });
}
```